### PR TITLE
Class API for Entities, Features and Labels

### PR DESCRIPTION
**Pull request by epps**
# Description

This PR introduces a new API for defining entities, features and labels.

### Select type(s) of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation update

# Checklist:

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] I have fixed any merge conflicts

### CODE REVIEW

File: client/src/featureform/cli.py [hunk 1]
```diff
@@ -231,7 +231,11 @@ def read_url(url):
 
 def exec_file(file, name):
     code = compile(file.read(), name, "exec")
-    exec(code)
+    # Create a new global namespace for each file to ensure that
+    # global variables, such as `ff`, are not undefined in the
+    # context of class attribute assignments (e.g. `label = ff.Label()`)
+    file_globals = {}
```
Comment (simba-git) [inline client/src/featureform/cli.py:237]: lol, I'm sorry you had to debug this, but great work

File: client/src/featureform/register.py [hunk 2]
```diff
@@ -802,9 +849,19 @@ def __call__(self, fn):
             if self.name is nv[0] and self.variant is nv[1]:
                 raise ValueError(f"Transformation cannot be input for itself: {self.name} {self.variant}")
         self.query = dill.dumps(fn.__code__)
-        fn.register_resources = self.register_resources
-        fn.name_variant = self.name_variant
-        return fn
+        # Given the function type does not implement __getitem__ we need to wrap it
```
Comment (simba-git) [inline client/src/featureform/register.py:852]: Noticed that you copy and pasted this, would it make sense to make it its own function or put the logic in the constructor?

File: client/src/featureform/register.py [hunk 3]
```diff
@@ -3947,6 +4010,194 @@ def search(self, raw_query, local=False):
...
+            entity_column=self.entity_column,
+            owner=self.owner,
+            inference_store=self.inference_store,
+            features=features,
+            labels=labels,
+            timestamp_column=self.timestamp_column,
+            schedule=self.schedule,
+        )
+    
+    def get_features_and_labels(self) -> Tuple[List[ColumnMapping], List[ColumnMapping]]:
```
Comment (simba-git) [inline client/src/featureform/register.py:4069]: would it make sense to separate these into two methods? If not would it make sense to drop the get_ in the name. Super minor nitpick feel free to ignore

File: client/src/featureform/register.py [hunk 4]
```diff
@@ -3947,6 +4010,194 @@ def search(self, raw_query, local=False):
...
+            if resource.variant != variant_key:
+                raise ValueError(
+                    f"Variant name {variant_key} does not match resource variant name {resource.variant}"
+                )
+
+    def register(self):
+        for resource in self.resources.values():
+            resource.register()
+
+# Class API
```
Comment (simba-git) [inline client/src/featureform/register.py:4108]: Would cut this comment

**Review by simba-git — DISMISSED** [review]
Looks great, nice work

Left some minor nitpicks that you can totally ignore as well

**Comment by aolfat** [comment]
I downloaded black formatting for my stuff: https://black.readthedocs.io/en/stable/. I've been using it and trying to get everyone else aboard.

File: client/src/featureform/register.py [hunk 5]
```diff
@@ -3947,6 +4010,194 @@ def search(self, raw_query, local=False):
             return search(processed_query, self._host)
 
 
+class ColumnResource:
```
Comment (aolfat) [inline client/src/featureform/register.py:4013]: You're effectively creating an abstract class here right? You can force the inability to instantiate with this -> https://www.geeksforgeeks.org/abstract-classes-in-python/

File: client/src/featureform/register.py [hunk 6]
```diff
@@ -3947,6 +4010,194 @@ def search(self, raw_query, local=False):
...
+    """
+    Base class for all column resources. This class is not meant to be instantiated directly.
```
Comment (aolfat) [inline client/src/featureform/register.py:4015]: > This class is not meant to be instantiated directly.

How about we make this an abstract class? https://www.geeksforgeeks.org/abstract-classes-in-python/

File: client/src/featureform/register.py (hunk 2 above)
Comment (epps) [reply client/src/featureform/register.py:852]: Definitely cleaner to wrap this up in the wrapper class. Thanks for the suggestion!

File: client/src/featureform/register.py (hunk 6 above)
Comment (epps) [reply client/src/featureform/register.py:4015]: I took a look at this as I was unfamiliar with the module, but I'm not sure it fits the use case here. I actually don't want to reimplement the methods in the derived classes, so having a functional base class is what I want, no?

File: client/tests/test_class_api.py [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+        custom_marks
+    )
+    label_entity_column = "CustomerID" if is_local else "customerid"
+
+    with pytest.raises(Exception, match="Expected 2 columns"):
+        source[[label_entity_column]]
+
+
+@pytest.fixture(autouse=True)
+def before_and_after_each(setup_teardown):
```
Comment (aolfat) [inline client/tests/test_class_api.py:119]: this has been copied pasta'd in a lot of places. I'm not sure what the correct solution is but ideally we don't' have to.

**Review by sdreyer — APPROVED** [review]

**Pull request merged**